cnn_comparator/simf_cnn_NCCL_parallel.py

```diff
@@ -0,0 +1,126 @@
+import os
+import torch
+import torch.distributed as dist
+import torch.nn as nn
+import torch.optim as optim
+from torch.utils.data import DataLoader, Subset
+from torchvision import datasets, transforms
+from libensemble.message_numbers import PERSIS_STOP, STOP_TAG, EVAL_SIM_TAG, WORKER_DONE
+
+# Define a Convolutional Neural Network
+class MNISTCNN(nn.Module):
+    def __init__(self):
+        super(MNISTCNN, self).__init__()
+        self.conv1 = nn.Conv2d(1, 32, kernel_size=3)  # 32 filters, 3x3 kernel
+        self.conv2 = nn.Conv2d(32, 64, kernel_size=3)  # 64 filters, 3x3 kernel
+        self.pool = nn.MaxPool2d(2)  # 2x2 pooling
+        self.fc1 = nn.Linear(64 * 12 * 12, 128)  # Flattened features to 128 units
+        self.fc2 = nn.Linear(128, 10)  # Output layer for 10 classes
+
+    def forward(self, x):
+        x = torch.relu(self.conv1(x))  # First convolution + activation
+        x = torch.relu(self.conv2(x))  # Second convolution + activation
+        x = self.pool(x)  # Pooling
+        x = x.view(-1, 64 * 12 * 12)  # Flatten the features
+        x = torch.relu(self.fc1(x))  # Fully connected layer
+        return self.fc2(x)  # Output layer
+
+
+def train(rank, group_size, epochs, batch_size, use_gpu):
+    # Redirect stdout to a log file for each rank
+    log_file = f"rank_{rank}_output.log"
+    with open(log_file, "w") as f:
+        os.dup2(f.fileno(), 1)  # Redirect stdout
+        os.dup2(f.fileno(), 2)  # Redirect stderr
+    # Initialize process group
+    dist.init_process_group("nccl")
+
+    # Toggle device based on use_gpu flag
+    device = torch.device(f"cuda:{rank % torch.cuda.device_count()}" if use_gpu else "cpu")
+
+    # Prepare dataset and DataLoader
+    transform = transforms.Compose([
+        transforms.ToTensor(),
+        transforms.Normalize((0.5,), (0.5,))
+    ])
+    dataset = datasets.MNIST(root="./data", train=True, transform=transform, download=False)
+    train_loader = DataLoader(dataset, batch_size=batch_size, shuffle=True, num_workers=8)
+    # Initialize model and optimizer
+    model = MNISTCNN().to(device)
+    optimizer = optim.SGD(model.parameters(), lr=0.01)
+
+    for epoch in range(epochs):
+        for batch_idx, (data, target) in enumerate(train_loader):
+            data, target = data.to(device), target.to(device)
+
+            optimizer.zero_grad()
+            output = model(data)
+            loss = nn.CrossEntropyLoss()(output, target)
+            loss.backward()
+
+            # Synchronize gradients across processes
+            for param in model.parameters():
+                dist.all_reduce(param.grad, op=dist.ReduceOp.SUM)
+                param.grad /= group_size  # SH averaging?
+
+            # SH TODO - consider inheriting dist (torch.distributed)
+            # and overriding dist.all_reduce() to do libE stuff
```

Comment:
torch.distributed is a module not a class, but could extract these lines to some provided function.